core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,241 @@
+package org.bykn.bosatsu
+
+import cats.data.{NonEmptyList, State}
+import com.stripe.dagon.Memoize
+import cats.Id
+import cats.implicits._
+
+sealed abstract class NormalExpression {
+
+}
+
+object NormalExpression {
+  case class App(fn: NormalExpression, arg: NormalExpression) extends NormalExpression
+  case class ExternalVar(pack: PackageName, defName: String) extends NormalExpression
+  case class Match(arg: NormalExpression, branches: NonEmptyList[(Int, NormalExpression)]) extends NormalExpression
+  case class LambdaVar(index: Int) extends NormalExpression
+  case class Lambda(expr: NormalExpression) extends NormalExpression
+  case class Struct(enum: Int, args: List[NormalExpression]) extends NormalExpression
+  case class Literal(lit: Lit) extends NormalExpression
```

Comment:
do we need a separate AST here or could we reuse `Expr[T]` with an appropriate `T` value?